src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -1715,6 +1715,117 @@ theorem toInt_sdiv (a b : BitVec w) : (a.sdiv b).toInt = (a.toInt.tdiv b.toInt).
 
+/-- Unsigned division is zero if and only if either the denominator is zero,
+or the numerator is unsigned less than the denominator -/
+@[simp]
+theorem udiv_eq_zero_iff_eq_zero_or_lt {x y : BitVec w} :
+     x / y = 0#w ↔ (y = 0#w ∨ x < y) := by
+  constructor
+  · intros h
+    obtain h := toNat_inj.mpr h
+    simp only [toNat_udiv, toNat_ofNat, zero_mod, Nat.div_eq_zero_iff] at h
+    rcases h with h | h
+    · left
+      apply eq_of_toNat_eq
+      simp [h]
+    · right
+      simp [lt_def, h]
+  · intros h
+    apply eq_of_toNat_eq
+    simp only [toNat_udiv, toNat_ofNat, zero_mod, Nat.div_eq_zero_iff]
+    simp only [lt_def] at h
+    rcases h with h | h <;> simp [h]
+
+@[simp]
+theorem udiv_ne_zero_iff_ne_zero_and_le {x y : BitVec w} :
+     x / y ≠ 0#w ↔ (y ≠ 0#w ∧ y ≤ x) := by
+  simp only [ne_eq, udiv_eq_zero_iff_eq_zero_or_lt, _root_.not_or, BitVec.not_lt]
+
+private theorem neg_udiv_eq_intmin_iff_eq_intmin_eq_one_of_msb_eq_true
+    {x y : BitVec w} (hx : x.msb = true) (hy : y.msb = false) :
+    -x / y = intMin w ↔ (x = intMin w ∧ y = 1#w) := by
+  constructor
+  · intros h
+    rcases w with _ | w; decide +revert
+    have : (-x / y).msb = true := by simp [h, msb_intMin];
+    rw [msb_udiv] at this
+    simp only [bool_to_prop] at this
+    obtain ⟨hx₂, hy⟩ := this
+    simp only [beq_iff_eq] at hy
+    subst hy
+    simp only [udiv_one, zero_lt_succ, neg_eq_intMin] at h
+    simp [h]
+  · rintro ⟨hx, hy⟩
+    subst hx
+    subst hy
+    simp
+
+theorem msb_sdiv_eq_decide {x y : BitVec w} :
+    (x.sdiv y).msb = ((0 < w) &&
+      -- x +ve, y +ve: never negative..
+      (!x.msb && y.msb && (-y ≤ x ∧ y ≠ 0#w)) || -- x +ve, y -ve: -ve when result nonzero.
+      (x.msb && !y.msb && (y ≤ -x && y ≠ 0#w)) || -- x -ve, y +ve: -ve when result nonzero.
+      (x.msb && y.msb && (x = intMin w && y = -1#w))) -- x -ve, y -ve:  `intMin / -1 = intMin` -ve.
+     := by
+  by_cases hw : w = 0; subst hw; decide +revert
+  simp only [show 0 < w by omega, decide_true, ne_eq, decide_and, decide_not, Bool.true_and,
+    sdiv_eq, udiv_eq]
+  rcases hxmsb : x.msb <;> rcases hymsb : y.msb
+  · -- x:false, y:false
+    simp [hxmsb, hymsb, msb_udiv_eq_false_of, Bool.not_false, Bool.and_false, Bool.false_and,
+      Bool.and_true, Bool.or_self, Bool.and_self]
+  · -- x:false, y:true
+    simp only [hxmsb, hymsb, msb_neg, msb_udiv_eq_false_of, bne_false, Bool.not_false,
+      Bool.and_self, ne_zero_of_msb_true, decide_false, Bool.and_true, Bool.true_and, Bool.not_true,
+      Bool.false_and, Bool.or_false, bool_to_prop]
+    have : x / -y ≠ intMin w := by
+      intros h
+      have : (x  / -y).msb = (intMin w).msb := by simp only [h]
+      simp only [msb_intMin, show 0 < w by omega, decide_true] at this
+      rw [msb_udiv] at this
+      simp only [and_eq_true, beq_iff_eq] at this
+      obtain ⟨hcontra, _⟩ := this
+      simp only [hcontra, true_eq_false] at hxmsb
```

Comment:
why not extracting another lemma like you did for `neg_udiv_eq_intmin_iff_eq_intmin_eq_one_of_msb_eq_true`? Also, I wonder if it would be possible to have a single, more general version of the lemma.